specs/capella/beacon-chain.md

```diff
@@ -323,9 +328,18 @@ def process_withdrawals(state: BeaconState, payload: ExecutionPayload) -> None:
     for expected_withdrawal, withdrawal in zip(expected_withdrawals, payload.withdrawals):
         assert withdrawal == expected_withdrawal
         decrease_balance(state, withdrawal.validator_index, withdrawal.amount)
-    if len(expected_withdrawals) > 0:
+
+    if len(expected_withdrawals) != 0:
         latest_withdrawal = expected_withdrawals[-1]
         state.next_withdrawal_index = WithdrawalIndex(latest_withdrawal.index + 1)
+
+    if len(expected_withdrawals) < MAX_WITHDRAWALS_PER_PAYLOAD:
+        # advance sweep if there were not a full set of withdrawals
+        next_index = state.next_withdrawal_validator_index + MAX_VALIDATORS_PER_WITHDRAWALS_SWEEP
+        next_validator_index = next_index % len(state.validators)
+        state.next_withdrawal_validator_index = ValidatorIndex(next_validator_index)
+    else:
+        latest_withdrawal = expected_withdrawals[-1]
         next_validator_index = ValidatorIndex((latest_withdrawal.validator_index + 1) % len(state.validators))
         state.next_withdrawal_validator_index = next_validator_index
```

Comment:
I can add more comments but we can't pull out the `latest_withdrawal = expected_withdrawals[-1]` assignment from its current place bc `len(expected_withdrawals) == 0` is possible which would raise an error

and there is no 'safe' built-in method for indexing on lists so there isn't some kind of pythonic thing we could do (which imo we don't want to actually do)

Reply:
hmm it isn't safe as currently written either though. The `else` isn't scoped by non-zero expected_withdrawals. will write an updated suggestion

Reply:
> hmm it isn't safe as currently written either though. The `else` isn't scoped by non-zero expected_withdrawals. will write an updated suggestion

It is, I'm fine with @ralexstokes' intent to explicitly assign in case people get confused by the scoping, but it's perfectly valid to not have the second `latest_withdrawal` assignment.
```shell
$cat danny.py
danny = 1
if 1 > 0:
    danny = 4
if 2 > 3:
    danny = 1
else:
    print(danny)

$ python danny.py
4
```
```

Reply:
okay, how about this. I really don't like relying on python's scoping rules. If not this, then I'd do two var assignments in different places 

```suggestion
    
    # Update the next withdrawal index if this block contained withdrawals
    if len(expected_withdrawals) != 0:
        state.next_withdrawal_index = WithdrawalIndex(expected_withdrawals[-1].index + 1)

    # Update the next validator index to start the next withdrawal sweep
    if len(expected_withdrawals) == MAX_WITHDRAWALS_PER_PAYLOAD:
        # Next sweep starts after the latest withdrawal's validator index
        next_validator_index = ValidatorIndex((expected_withdrawals[-1].validator_index + 1) % len(state.validators))
        state.next_withdrawal_validator_index = next_validator_index
    else:
        # Advance sweep by the max length of the sweep if there was not a full set of withdrawals
        next_index = state.next_withdrawal_validator_index + MAX_VALIDATORS_PER_WITHDRAWALS_SWEEP
        next_validator_index = next_index % len(state.validators)
        state.next_withdrawal_validator_index = ValidatorIndex(next_validator_index)
```

Reply:
this seems fine, I've made the update